activerecord/lib/active_record/autosave_association.rb

```diff
@@ -371,6 +372,11 @@ def before_save_collection_association
         true
       end
 
+      def after_save_collection_association
+        @new_record_before_save = false
+        true
```

Comment:
On the other hand, that has nothing to do with after_save callbacks, only before_save.  I'll try removing it and see what happens.